yarn-project/circuit-types/src/body.ts

```diff
@@ -0,0 +1,200 @@
+
+    const numberOfTxsIncludingEmpty = newNullifiers.length / MAX_NEW_NULLIFIERS_PER_TX;
+
+    for (let i = 0; i < numberOfTxsIncludingEmpty; i += 1) {
+      txEffects.push(
+        new TxEffect(
+          newNoteHashes.slice(i * MAX_NEW_COMMITMENTS_PER_TX, (i + 1) * MAX_NEW_COMMITMENTS_PER_TX),
+          newNullifiers.slice(i * MAX_NEW_NULLIFIERS_PER_TX, (i + 1) * MAX_NEW_NULLIFIERS_PER_TX),
+          newL2ToL1Msgs.slice(i * MAX_NEW_L2_TO_L1_MSGS_PER_TX, (i + 1) * MAX_NEW_L2_TO_L1_MSGS_PER_TX),
+          newPublicDataWrites.slice(
+            i * MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX,
+            (i + 1) * MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX,
+          ),
+          newContracts.slice(i * MAX_NEW_CONTRACTS_PER_TX, (i + 1) * MAX_NEW_CONTRACTS_PER_TX),
+          newContractData.slice(i * MAX_NEW_CONTRACTS_PER_TX, (i + 1) * MAX_NEW_CONTRACTS_PER_TX),
+          new TxEffectLogs(newEncryptedLogs!.txLogs[i], newUnencryptedLogs!.txLogs[i]),
+        ),
+      );
+    }
+
+    return new this(newL1ToL2Messages, txEffects);
+  }
+
+  /**
+   * Computes the calldata hash for the L2 block
+   * This calldata hash is also computed by the rollup contract when the block is submitted,
+   * and inside the circuit, it is part of the public inputs.
+   * @returns The calldata hash.
+   */
+  getCalldataHash() {
+    this.assertLogsAttached();
+
+    const computeRoot = (leafs: Buffer[]): Buffer => {
+      const layers: Buffer[][] = [leafs];
+      let activeLayer = 0;
+
+      while (layers[activeLayer].length > 1) {
+        const layer: Buffer[] = [];
+        const layerLength = layers[activeLayer].length;
+
+        for (let i = 0; i < layerLength; i += 2) {
+          const left = layers[activeLayer][i];
+          const right = layers[activeLayer][i + 1];
+
+          layer.push(sha256(Buffer.concat([left, right])));
+        }
+
+        layers.push(layer);
+        activeLayer++;
+      }
+
+      return layers[layers.length - 1][0];
+    };
+
+    const leafs: Buffer[] = this.txEffects.map(txEffect => txEffect.hash());
+
+    return computeRoot(leafs);
+  }
+
+  public assertLogsAttached() {
+    if (!this.areLogsAttached()) {
+      throw new Error('newEncryptedLogs and newUnencryptedLogs must be defined');
+    }
+  }
+
+  public areLogsAttached() {
+    return this.txEffects.every(txEffect => txEffect.logs !== undefined);
+  }
+
+  public areLogsEqual(encryptedLogs: L2BlockL2Logs, unencryptedLogs: L2BlockL2Logs) {
+    if (
+      this.txEffects.every(
+        (txEffect, i) =>
+          txEffect.logs?.encryptedLogs.equals(encryptedLogs.txLogs[i]) &&
+          txEffect.logs?.unencryptedLogs.equals(unencryptedLogs.txLogs[i]),
+      )
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  get encryptedLogs(): L2BlockL2Logs {
+    const logs = this.txEffects.map(txEffect => txEffect.logs!.encryptedLogs);
+
+    if (logs.length !== this.txEffects.length || logs.some(log => log === undefined)) {
```

Comment:
This check seems redundant if you prohibit undefined logs on tx effects (which should be done given that you nuked the log attaching).

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/pQjA8LkugIeLVkPzHKRq/c2510907-1838-40b6-aca3-7fd97c62dc32.png)